dandiapi/api/models/asset.py

```diff
@@ -125,25 +126,57 @@ def copy(cls, asset):
         return Asset(path=asset.path, blob=asset.blob, metadata=asset.metadata)
 
     @classmethod
-    def get_path(cls, path_prefix: str, qs: List[str]) -> Set:
+    def get_path(cls, dandiset_pk: str, version: str, path_prefix: str, qs: List[Asset]) -> Set:
         """
         Return the unique files/directories that directly reside under the specified path.
 
         The specified path must be a folder (must end with a slash).
         """
-        if not path_prefix:
-            path_prefix = '/'
         prefix_parts = [part for part in path_prefix.split('/') if part]
-        paths = set()
+        paths = {'folders': {}, 'items': {}}
+
+        # Declare base api url and kwargs for all asset reverse lookups
+        base_api_url = 'https://api.dandiarchive.org'
+        kwargs = {'versions__dandiset__pk': dandiset_pk, 'versions__version': version}
+
         for asset in qs:
-            path_parts = [part for part in asset['path'].split('/') if part]
+            path_parts = [part for part in asset.path.split('/') if part]
 
             # Pivot index is -1 (include all path parts) if prefix is '/'
             pivot_index = path_parts.index(prefix_parts[-1]) if len(prefix_parts) else -1
             base_path, *remainder = path_parts[pivot_index + 1 :]
-            paths.add(f'{base_path}/' if len(remainder) else base_path)
-
-        return sorted(paths)
+            is_folder = len(remainder)
+
+            if is_folder:
+                if paths['folders'].get(base_path) is None:
+                    paths['folders'][base_path] = {
+                        'path': f'{base_path}/',
+                        'size': 0,
+                        'num_items': 0,
+                        'created': asset.created,
+                        'modified': asset.modified,
+                    }
+
+                entry = paths['folders'][base_path]
+                entry['size'] += asset.size
+                entry['num_items'] += 1
+                entry['created'] = min(entry['created'], asset.modified)  # earlist
+                entry['modified'] = max(entry['modified'], asset.modified)  # latest
+            else:
+                kwargs['asset_id'] = asset.asset_id
+                asset_delete_url = base_api_url + reverse('asset-detail', kwargs=kwargs)
+                asset_download_url = base_api_url + reverse('asset-download', kwargs=kwargs)
```

Comment:
by the looks of it that `reverse` looks not that "speedy".  I think it would be well worth comparing against straight assignment of a URL based on our knowledge of what it should look like given asset information we have (or is anything lacking in the `Asset`?).

Reply:
The main idea is to prevent the need to change this URL, if we decide to change where the `asset-detail` and `asset-download` endpoints are mounted in the future. While that may not be likely to happen, it's good software practice.

I don't believe `reverse` is costly anyway, as it should just be little more than a dict lookup. I tested this locally by replacing the calls to `reverse` with a hardcoded string, and the results were identical (roughly ~180-200ms for a folder with 8 items on my local environment).

Reply:
Thanks for timing! sounds oddly long though if it is milliseconds for just 8 items: most likely there is overall overhead bias and actual loop IMHO should take well under 1ms per asset to consider AFAIK.  So might be worth timing by explicit time printouts as diff in time() before/after the loop for adequate comparison.  If it was in µs (AKA us) - then ok ;)

Reply:
Sorry I gave the wrong number, the 200ms number was for all the assets in the dandiset (which is a local copy of dandiset `000006`, so 53 assets). I believe the number for the folder with 8 assets was ~35ms.

Reply:
Regarding the timing: I was timing the asset loop only, not the whole request, done with python's `time.time()`.

Reply:
Thank you! Still too high if can't get optimized, since if I do math right would be about 4 sec for one with 1000 assets.  https://gui.dandiarchive.org/#/dandiset/000026 has 1225 assets ATM .

Reply:
But I refuse to believe that python is that slow!!! There must be a gotcha. BTW py-spy is quite nice if you do get multi seconds runs and just py-spy top the process of even create a flame graph ;-)

Reply:
@yarikoptic Are you okay with merging this as is and working on optimization later (to preserve momentum)? I still have to make the client-side changes however, so this merge might still be delayed anyway.

Reply:
I will leave it at your discretion since it seems it might have detrimental effect on timings significantly effecting top level directories (typically with folders) and I would have first tried to figure out if such approach is pragmatically feasible.

Also -- Don't we auto-deploy master -- and then it would break web UI since if I got it right it changes the output for `/files` endpoint, and should instead be repositioned against `staging` anyways so then both dandi-api and client(s) could then be rolled out at the same time?

Reply:
You're correct, we'd have to coordinate the client changes and this one, which is why this change might still be delayed. Staging is in a bit of a weird spot right now, so I'm not sure this PR will be able to take advantage of it.

Reply:
That is-partially-why I was looking to add this extended listing as an option, without changing the default in #292 so it could have been merged/tried/benchmarked/improved (and possibly discarded ;))  without requiring any change to web UI.
Anyways -- I am still intrigued by the slow performance.  If your timings hold, I am not sure that this would be a viable solution worth pursuing since it would be of great negative effect on web ui navigation of majority of the dandisets.
Gimme an hour -- I want to see if it is indeed Python which inflicts itself upon us here or it is something of django so we are least get the lower bound.  My plan -- take assets records from e.g. 000026 and see how long it would take to get listing on top level (without talking to django to compose URLs).

Reply:
wouldn't this be only for files that are viewable at a level? so this should never be run on all assets ever right?

Reply:
it would loop through all assets if ran for top level prefix `/`, but indeed would not need to mint urls for them etc.

<details>
<summary>With this dummy script where I copy pasted get_path as is and run it across metadata records with no-op "reverse" to get URL</summary> 

```shell
from typing import Dict, List

class Asset(dict):
    def __getattr__(self, item):
        return self.__getitem__(item)


def reverse(*args, **kwargs):
    # noop for now
    return "TODO"


def get_path(dandiset_pk: str, version: str, path_prefix: str, qs: List[Asset]) -> Dict:
    """
    Return the unique files/directories that directly reside under the specified path.

    The specified path must be a folder (must end with a slash).
    """
    # Declare base api url and kwargs for all asset reverse lookups
    base_api_url = 'https://api.dandiarchive.org'
    kwargs = {'versions__dandiset__pk': dandiset_pk, 'versions__version': version}

    # Initialize dicts to store folders and files
    folders = {}
    files = {}

    for asset in qs:
        # Get the remainder of the path after path_prefix
        base_path: str = asset.path[len(path_prefix):].strip('/')

        # Since we stripped slashes, any remaining slashes indicate a folder
        folder_index = base_path.find('/')
        is_folder = folder_index >= 0

        if is_folder:
            base_path = base_path[:folder_index]
            entry = folders.get(base_path)
            if entry is None:
                folders[base_path] = {
                    'size': asset.size,
                    'num_files': 1,
                    'created': asset.created,
                    'modified': asset.modified,
                }
            else:
                entry['size'] += asset.size
                entry['num_files'] += 1
                entry['created'] = min(entry['created'], asset.created)  # earliest
                entry['modified'] = max(entry['modified'], asset.modified)  # latest
        else:
            kwargs['asset_id'] = asset.asset_id
            asset_delete_url = base_api_url + reverse('asset-detail', kwargs=kwargs)
            asset_download_url = base_api_url + reverse('asset-download', kwargs=kwargs)

            files[base_path] = {
                'size': asset.size,
                'created': asset.created,
                'modified': asset.modified,
                'delete_url': asset_delete_url,
                'download_url': asset_download_url,
            }

    return {
        'folders': folders,
        'files': files,
    }



import yaml
import time
# produced by dandi ls -r -f yaml DANDI:000026 > 000026-dandi-ls-r.json
# to not mess with pagination , see https://github.com/dandi/dandi-api/issues/252
# and overcome https://github.com/dandi/dandi-cli/issues/652 for -f json_pp
with open('000026-dandi-ls-r.yaml') as f:
    # wrap them into that odd Asset to provide `.__getattr__` so we could use original code as is
    assets = [Asset(_) for _ in yaml.load(f, Loader=yaml.SafeLoader) if 'asset_id' in _]

print(f"Read {len(assets)} assets")
"""noted a bug of eating first letter
*(Pdb) out['folders']
{'erivatives': {'size': 62098107735, 'num_files': 83, 'created': '2021-04-21T08:45:29.941149Z', 'modified': '2021-05-18T13:13:45.601444Z'}, 'awdata': {'size': 836971584508, 'num_files': 1142, 'created': '2021-04-21T08:49:14.803490Z', 'modified': '2021-05-18T13:15:42.881510Z'}}
"""
for prefix in ('/', 'derivatives/', 'rawdata/sub-I46/ses-SPIM/'):
    t0 = time.time()
    out = get_path("000026", "draft", prefix, assets)
    dt = time.time() - t0
    print(f"{prefix}: Got {len(out['folders'])} folders and {len(out['files'])} files in {dt:.3f} sec")
```
</details>

I got

1. timing -- Python is FAST enough! ;-) (despite my ugly `__getattr__`)
```
$> python time_api_get_path.py
Read 1225 assets
/: Got 2 folders and 0 in 0.002 sec
derivatives/: Got 9 folders and 0 in 0.002 sec
rawdata/sub-I46/ses-SPIM/: Got 4 folders and 1154 in 0.003 sec
```

so the loop/Python is not the issue but something about using `__getattr__` on assets or `reverse` which slows it down.

2. my recommended "simple" way logic might benefit from "guarding" the assumption that asset path might starts with `/` -- I do not think it would anyhow affect timing but would make cost more robust (talking about assumptions on what outside code provides), since otherwise it might  eat up leading character:
```
print(get_path("000026", "draft", "/", assets))
```
results in
```
{'folders': {'erivatives': {'size': 62098107735, 'num_files': 83, 'created': '2021-04-21T08:45:29.941149Z', 'modified': '2021-05-18T13:13:45.601444Z'}, 'awdata': {'size': 836971584508, 'num_files': 1142, 'created': '2021-04-21T08:49:14.803490Z', 'modified': '2021-05-18T13:15:42.881510Z'}}, 'files': {}}
```
since currently we do not prefix asset paths with `/` and we I do not think we need to revert that or make it special.
```
$> curl --silent -X GET "https://api.dandiarchive.org/api/dandisets/000026/versions/draft/assets/?page=1&page_size=100000" -H "accept: application/json" | jq '.' | grep path | head
      "path": "derivatives/EPIC/dataset_description.json",
      "path": "derivatives/EPIC/sub-I46/sub-I46_ses-MRI_flip-1_VFA.json",
      "path": "derivatives/EPIC/sub-I46/sub-I46_ses-MRI_flip-2_VFA.json",
      "path": "derivatives/EPIC/sub-I48/sub-I48_ses-MRI_flip-1_VFA.json",
```

Reply:
I did already test removing `reverse` and it seemed to have no effect at all. However, the main reason for the slowness may have been how we were accessing the asset blobs. Since we look at the blob to check the size, we need to get that from the database as well. With what you just tested, those were two separate calls. So we would make a call to the database to get all relevant assets, and the for _each_ individual asset, we'd make a call to the database to get that blob, just for the size.

In 3780cb2 I've fixed that, so it fetches the blobs at the same time as the asset. This seems to show a ~6-7x speed improvement.

Reply:
for my own education, that `self.get_queryset().select_related('blob')` in 3780cb2 is like a joint between assets and blobs? If that was the only reason -- surprised we are not getting even bigger boost.  
Well, may be now the `reverse` is the one which is taking most of the time and you would spot the effect (i.e. if before size was the major factor, it would have masked any effect of `reverse`)?

Reply:
note: for the top level of dandiset, since we do not have files there, `reverse` should have no effect.  So if it is still is taking `ms`nds "per each asset" -- there must be some additional culprit in how data is accessed which might need to be ironed out.

Reply:
> for my own education, that `self.get_queryset().select_related('blob')` in [3780cb2](https://github.com/dandi/dandi-api/commit/3780cb2073dfc04713c98a1b659ddb12a73c2ee1) is like a joint between assets and blobs?

Yes.


> Well, may be now the `reverse` is the one which is taking most of the time and you would spot the effect (i.e. if before size was the major factor, it would have masked any effect of `reverse`)?

I tried the same experiment again, and it still seems that removing `reverse` has no effect. On my local setup, it's taking ~100-120ms at the top level, with 53 assets. It appears to also take roughly the same time when in a folder with 8 assets, with or without the `reverse` calls.

Reply:
I should point out, the above times are for the entire request. If I just time the asset loop, it's pretty consistently `10ms` for all assets. So quite a reduction from the previous `200ms` for all assets.

Reply:
>  it's pretty consistently 10ms for all assets

regardless of the number of them (8 vs 53)? and that is done not for top level directory where there is no assets at that level (so `reverse` would not be even triggered), what about the folder with the assets?  

anyways -- it seems we would be shooting in the dark by simply rolling it out straight into deployment -- what was the problem with the staging besides flooding sentry? could we disconnect sentry ? then we could have populated some dummy datasets with some good number of assets (I could even mimic real ones, just make them with dummy small blobs so we do not torture that bucket) to assess the performance "first hand".

Reply:
I did some more investigation, and it seems that most of the time in that loop (the `10ms`) is spent on database overhead. If I cast the query to a list beforehand (to force the query to run before the loop), the loop takes roughly 0.2 milliseconds for all assets, and roughly _`1ms`_ for the folder with 8 assets. So it does seem that reverse is having an impact (testing without the reverse call, the loop takes `0.03ms`, which is expected.

If I time the casting of the queryset to a list, that call takes ~`10ms`. This would explain why we were seeing similar request times for 53 vs 8 assets. I'd guess the rest of the request time (total request taking a little over `100ms`) is spent handling the request. 

I still think it'd be good to test this in staging with large datasets, but with this new information I'm more confident that this implementation scales well to larger datasets. The calls to `reverse` are definitely taking up some time, but I think we should test out how it actually performs on large datasets before jumping the gun and doing away with it. If we test it out and see that the performance really is an issue, then I'd say we should use a different approach.

Reply:
Thanks for the investigation!  
- indeed sounds like we could proceed forward.  if `reverse` would become prohibitively expensive, we could provide ad-hoc composition of the urls as I thought about originally
- we probably shouldn't merge it straight into master at this point. With all the vacationing upon many, we agreed in today's brief standup call that bringing up staging so we could test + orchestrate joint releasing of dandi-api + web-ui would be the way to go forward (@waxlamp please attest that I am I not mis-interpreting our verbal decision).

So - may be this PR could be "rebased" against staging and merged meanwhile with accompanying changes to web-ui (dandiarchive) also merged into its own staging?

Reply:
I'm reviewing this PR now, so just to resolve some threads here -

> * indeed sounds like we could proceed forward.  if `reverse` would become prohibitively expensive, we could provide ad-hoc composition of the urls as I thought about originally

Agreed. Right now the slow loading time for assets of large datasets is an issue in the client, specifically the loading time is dominated by the fact that "n" HTTP requests need to be made to the server for "n" folders/files. Once the web client is updated to make use of the changes in this PR we can see if `reverse` is still a problem.

> * we probably shouldn't merge it straight into master at this point. With all the vacationing upon many, we agreed in today's brief standup call that bringing up staging so we could test + orchestrate joint releasing of dandi-api + web-ui would be the way to go forward (@waxlamp please attest that I am I not mis-interpreting our verbal decision).
So - may be this PR could be "rebased" against staging and merged meanwhile with accompanying changes to web-ui (dandiarchive) also merged into its own staging?

Once we have staging properly set up (should be soon), we can just merge this straight to master and test the changes using our staging infrastructure.

Reply:
> Once we have staging properly set up (should be soon), we can just merge this straight to master and test the changes using our staging infrastructure.

Just to provide extra context: Mike and Daniel are working now to arrange the deployment infrastructure so that `master` autodeploys to staging, and creating a release causes a deployment to production. This lets us test all changes comprehensively in staging, and gives us fine control over when to push those changes to production. (I believe this is the approach outlined in a design doc.)

Reply:
confirming -- matches description in https://github.com/dandi/dandi-api/blob/master/doc/design/staging-1.md#release-procedure ... I guess eventually we might need to elaborate on having `maint` or alike (could be throw away release branches) branch for "hot releases" so we do not need to release "staging" `master` for some quick fix (e.g. like a "fix" for #384)